dbms/src/Columns/IColumn.h

```diff
@@ -262,39 +270,43 @@ class IColumn : public COWPtr<IColumn>
     /// Note:
     /// 1. The pos.size() must be greater than or equal to length.
     /// 2. If has_null is true, then the pos[i] could be nullptr, which means the i-th element does not need to be serialized.
-    virtual void serializeToPosForCmp(
+    virtual void serializeToPos(
         PaddedPODArray<char *> & /* pos */,
         size_t /* start */,
         size_t /* length */,
-        bool /* has_null */,
-        const TiDB::TiDBCollatorPtr & /* collator */,
-        String * /* sort_key_container */) const
+        bool /* has_null */) const
         = 0;
-    virtual void serializeToPos(
+    /// Similar to serializeToPos, but there are two changes to make sure compare semantics is kept:
+    /// 1. For ColumnString with collator, this method first decode collator and then serialize to pos.
+    /// 2. For ColumnNullable(ColumnXXX), a default value of the nested column will be serialized if this row is null.
+    virtual void serializeToPosForCmp(
         PaddedPODArray<char *> & /* pos */,
         size_t /* start */,
         size_t /* length */,
-        bool /* has_null */) const
+        const NullMap * /*nullmap*/,
```

Comment:
Why remove `has_null`? `has_null` means the `pos` array may have some null elements. `nullmap` means there is a `ColumnNullable` outside of this nested column. They have different purposes.

Reply:
no one use `has_null`